src/any_agent/frameworks/llama_index.py

```diff
@@ -1,35 +1,501 @@
+import json
+import uuid
+from collections.abc import Sequence
+from json import JSONDecodeError
 from typing import TYPE_CHECKING, Any, cast
 
-from litellm import acompletion
-from litellm.utils import supports_response_schema
-from pydantic import BaseModel, ValidationError
+from any_llm import AnyLLM, acompletion
+from any_llm.types.completion import (
+    ChatCompletion,
+    ChatCompletionChunk,
+    ChatCompletionMessage,
+)
+from pydantic import BaseModel, Field, PrivateAttr, ValidationError
 
 from any_agent import AgentConfig, AgentFramework
 from any_agent.logging import logger
 from any_agent.tools.final_output import prepare_final_output
+from any_agent.vendor.llama_index_utils import (
+    force_single_tool_call,
+    from_openai_message_dict,
+    to_openai_message_dicts,
+    update_tool_calls,
+)
 
 from .any_agent import AnyAgent
 
+if TYPE_CHECKING:
+    from collections.abc import AsyncIterator, Iterator
+
+    from llama_index.core.agent.workflow.workflow_events import AgentOutput
+    from llama_index.core.llms import LLM
+
 try:
     from llama_index.core.agent.workflow import (
         BaseWorkflowAgent,
         FunctionAgent,
         ReActAgent,
     )
-    from llama_index.llms.litellm import LiteLLM
+    from llama_index.core.base.llms.generic_utils import (
+        achat_to_completion_decorator,
+        astream_chat_to_completion_decorator,
+        chat_to_completion_decorator,
+        stream_chat_to_completion_decorator,
+    )
+    from llama_index.core.base.llms.types import (
+        ChatMessage,
+        ChatResponse,
+        ChatResponseAsyncGen,
+        ChatResponseGen,
+        CompletionResponse,
+        CompletionResponseAsyncGen,
+        CompletionResponseGen,
+        LLMMetadata,
+        MessageRole,
+    )
+    from llama_index.core.constants import DEFAULT_TEMPERATURE
+    from llama_index.core.llms.callbacks import (
+        llm_chat_callback,
+        llm_completion_callback,
+    )
+    from llama_index.core.llms.function_calling import FunctionCallingLLM
+    from llama_index.core.tools import ToolSelection
+    from llama_index.core.tools.types import BaseTool
 
     DEFAULT_AGENT_TYPE = FunctionAgent
-    DEFAULT_MODEL_TYPE = LiteLLM
+
+    class AnyLLMWrapper(FunctionCallingLLM):
```

Comment:
Why is the class definition inside the `try:` block?

Reply:
Since the FunctionCallingLLM is a llama_index class, the class definition depends on the llama_index package existing, so if this was outside the try/catch it would throw name errors and would bypass our llama_index_available logic. Which is now making me think that we may need to go back and touch this up for the other providers, not sure we did this consistently in the conversion process.